File: 34-find-first-and-last-position-of-element-in-sorted-array-medium/solution.py
from typing import List

class Solution:
    @staticmethod
    def find_first_occurance(nums: list, target: int)->int:
        """
        Find the index of the first occurance of target in a list using binary search
        """
        start = 0
        end = len(nums) - 1
        middle = int(((start + end) - ((start + end) % 2)) / 2)

        while start < end:
            # if m > target => move end and ignore m
            if nums[middle] > target:
                end = middle - 1
            # else if m < target => move start and ignore m
            elif nums[middle] < target:
                start = middle + 1
            # else => move end but DO NOT ignore m
            else:
                end = middle

            # recalculate m
            middle = int(((start + end) - ((start + end) % 2)) / 2)


        return middle if nums[middle] == target else -1



    @staticmethod
    def find_last_occurance(nums: list, target: int)->int:
        """
        Find the index of the first occurance of target in a list using binary search
        """
        start = 0
        end = len(nums) - 1
        middle = int(((start + end) + ((start + end) % 2)) / 2)

        while start < end:
            # if m > target => move end and ignore m
            if nums[middle] > target:
                end = middle - 1
            # else if m < target => move start and ignore m
            elif nums[middle] < target:
                start = middle + 1
            # else => move start but DO NOT ignore m
            else:
                start = middle

            # recalculate m
            middle = int(((start + end) + ((start + end) % 2)) / 2)


        return middle if nums[middle] == target else -1


    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0 or nums[0] > target or nums[len(nums) - 1] < target:
            return [-1, -1]
        return [self.find_first_occurance(nums, target), self.find_last_occurance(nums, target)]
        

File: 34-find-first-and-last-position-of-element-in-sorted-array-medium/solution.ts
export function findFirstOccurance(nums: number[], target: number) {    
    let start = 0;
    let end = nums.length - 1;

    if (nums[end] < target || nums[start] > target) return -1;
    
    // floor the middle
    let first = ((start + end) - ((start + end) % 2)) / 2;
    
    while (start < end) {
        if (nums[first] > target) end = first - 1;
        else if (nums[first] < target) start = first + 1;
        else end = first
        
        first = ((start + end) - ((start + end) % 2)) / 2;
    }

    return nums[first] === target ? first : -1;
}


function findLastOccurance(nums, target) {
    let start = 0;
    let end = nums.length - 1;
    
    if (nums[end] < target || nums[start] > target) return -1;
    
    // ceil the middle
    let last = ((start + end) + ((start + end) % 2)) / 2;
    
    while (start < end) {
        if (nums[last] > target) end = last - 1;
        else if (nums[last] < target) start = last + 1;
        else start = last
        
        last = ((start + end) + ((start + end) % 2)) / 2;
    }
    
    return nums[last] === target ? last : -1;
 }


function searchRange(nums, target) {
    return [findFirstOccurance(nums, target), findLastOccurance(nums, target)];
};